Framework/Geometry/inc/MantidGeometry/Objects/BeamlineRayTracer.h

```diff
@@ -0,0 +1,335 @@
+  Track probeTrack(trkStart, trkDirection);
+  int intercepts =
+      componentInfo.shape(componentIndex).interceptSurface(probeTrack);
+
+  // Iterate over track
+  Track::LType::const_iterator it;
+  for (it = probeTrack.cbegin(); it != probeTrack.cend(); ++it) {
+    // Starting position
+    V3D in = it->entryPoint;
+    componentInfo.rotation(componentIndex).rotate(in);
+
+    // Use the scale factor
+    in *= componentInfo.scaleFactor(componentIndex);
+    in += componentInfo.position(componentIndex);
+    V3D out = it->exitPoint;
+    componentInfo.rotation(componentIndex).rotate(out);
+
+    // Use the scale factor
+    out *= componentInfo.scaleFactor(componentIndex);
+    out += componentInfo.position(componentIndex);
+
+    // Create a link and add it to the track
+    track.addLink(in, out, out.distance(track.startPoint()),
+                  componentInfo.shape(componentIndex),
+                  componentInfo.componentID(componentIndex)->getComponentID());
+  }
+}
+} // namespace IntersectionHelpers
+
+/**
+ * Fire the test ray at the instrument and perform a bread-first search of the
+ * object tree to find the objects that were intersected.
+ *
+ * @param testRay :: An input/output parameter that defines the track and
+ * accumulates the intersection results.
+ * @param componentInfo :: The object that will provide access to component
+ * information.
+ */
+void fireRay(Track &track, const ComponentInfo &componentInfo) {
+  // Cast size of componentInfo to int
+  int size = static_cast<int>(componentInfo.size());
+  --size;
+
+  // Loop through the bounding boxes in reverse
+  // (essentially a breadth first search)
+  for (int i = size; i >= 0; --i) {
+    // Store the bounding box
+    BoundingBox box = componentInfo.boundingBox(i);
+
+    // Test for intersection
+    if (box.doesLineIntersect(track)) {
+
+      // Store the type of the child component
+      auto childComponentType = componentInfo.componentType(i);
+      // Store the type of the child's grandparent
+      auto grandParent = componentInfo.componentType(
+          componentInfo.parent(componentInfo.parent(i)));
+
+      // Don't want to count the bank and the detector
+      if (grandParent == Types::Rectangular &&
+          childComponentType == Types::Detector) {
+        continue;
+      }
+
+      // Process a rectangular bank or normal component
+      if (childComponentType == Types::Rectangular) {
+        IntersectionHelpers::checkIntersectionWithRectangularBank(
+            track, componentInfo, i);
+      } else {
+        IntersectionHelpers::checkIntersectionWithComponent(track,
+                                                            componentInfo, i);
+      }
+    }
+  }
+}
+
+/**
+ * Return the results of any trace() calls since the last call to getResults.
+ *
+ * @return A collection of links defining intersection information
+ */
+Links getResults(Track &resultsTrack) {
```

Comment:
Because everything is in a single header, you are not directing (restricting) client code to what they should be calling from all these methods. Everything is accessible. I think you only want `traceFromSource` to be declared in the header.